sklearn/utils/_encode.py

```diff
@@ -118,26 +153,57 @@ def _check_unknown(values, known_values, return_mask=False):
         Additionally returned if ``return_mask=True``.
 
     """
-    if values.dtype == object:
+    valid_mask = None
+
+    if values.dtype.kind in 'UO':
         uniques_set = set(known_values)
-        diff = list(set(values) - uniques_set)
+        diff = set(values) - uniques_set
         if return_mask:
             if diff:
                 valid_mask = np.array([val in uniques_set for val in values])
             else:
                 valid_mask = np.ones(len(values), dtype=bool)
-            return diff, valid_mask
+
+        none_in_diff = None in diff
+        nan_in_diff = np.nan in diff
+
+        if none_in_diff and nan_in_diff:
+            raise ValueError("Input wiith both types of missing, None and "
+                             "np.nan, is not supported")
+        if none_in_diff:
```

Comment:
Why is this necessary? Does ``sort`` fail to sort? Can you add a comment?

Reply:
This comment may have lost its context. The `nans` were removed from the sets above because they do not work with set differencing in this context.

As for the `Nones`, we could have left the `Nones` in the set, but it would complicate the `_extract_missing` helper.